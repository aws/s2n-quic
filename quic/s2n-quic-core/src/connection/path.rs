//! This module contains the Path implementation

use crate::{connection::ConnectionId, inet::SocketAddress};

/// Maintain metrics for each connection
#[derive(Debug, Copy, Clone)]
pub struct ConnectionMetrics {
    pub bytes_sent: usize,
    pub bytes_recv: usize,
}

impl Default for ConnectionMetrics {
    fn default() -> Self {
        ConnectionMetrics {
            bytes_sent: 0,
            bytes_recv: 0,
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Path {
    /// The peer's socket address
    pub peer_socket_address: SocketAddress,
    /// The connection id of the peer
    pub source_connection_id: ConnectionId,
    /// The the connection id the peer wanted to access
    pub destination_connection_id: ConnectionId,
    /// Holds metrics about bytes sent to the peer and received from the peer
    metrics: ConnectionMetrics,
    /// Tracks whether this path has passed Address or Path validation
    validated: bool,
}

/// A Path holds the local and peer socket addresses, conneciton ids, and metrics. It can be
/// validated or not-yet-validated.
impl Path {
    pub fn new(
        destination_connection_id: ConnectionId,
        peer_socket_address: SocketAddress,
        source_connection_id: ConnectionId,
    ) -> Self {
        Path {
            peer_socket_address,
            source_connection_id,
            destination_connection_id,
            metrics: Default::default(),
            validated: false,
        }
    }

    /// Called when bytes have been transmitted on this path
    pub fn on_bytes_transmitted(&mut self, bytes: usize) {
        self.metrics.bytes_sent += bytes;
    }

    /// Called when bytes have been received on this path
    pub fn on_bytes_received(&mut self, bytes: usize) {
        self.metrics.bytes_recv += bytes;
    }

    /// Returns whether this path has passed address validation
    pub fn is_validated(&self) -> bool {
        self.validated
    }

    //= https://tools.ietf.org/html/draft-ietf-quic-transport-29#section-8.1
    //# Prior to validating the client address, servers MUST NOT send more
    //# than three times as many bytes as the number of bytes they have
    //# received.
    pub fn mtu(&self, requested_size: usize) -> usize {
        if self.validated {
            return requested_size;
        }

        let limit = (self.metrics.bytes_recv * 3) - self.metrics.bytes_sent;
        if limit < requested_size {
            limit
        } else {
            requested_size
        }
    }

    /// Returns whether this path is blocked from transmitting more data
    pub fn at_amplification_limit(&self) -> bool {
        self.mtu(1) == 0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn amplification_limit_test() {
        let mut path = Path::new(
            ConnectionId::try_from_bytes(&[]).unwrap(),
            SocketAddress::default(),
            ConnectionId::try_from_bytes(&[]).unwrap(),
        );

        path.on_bytes_received(3);
        path.on_bytes_transmitted(8);
        assert_eq!(path.at_amplification_limit(), false);

        path.on_bytes_transmitted(1);
        assert_eq!(path.at_amplification_limit(), true);

        path.validated = true;
        // Validated paths should always be able to transmit
        assert_eq!(path.at_amplification_limit(), false);
    }

    #[test]
    fn mtu_test() {
        // TODO this would work better as a fuzz test
        let mut path = Path::new(
            ConnectionId::try_from_bytes(&[]).unwrap(),
            SocketAddress::default(),
            ConnectionId::try_from_bytes(&[]).unwrap(),
        );

        path.on_bytes_received(3);
        path.on_bytes_transmitted(8);

        // Verify we can transmit one more byte
        assert_eq!(path.mtu(1), 1);
        assert_eq!(path.mtu(10), 1);

        path.on_bytes_transmitted(1);
        // Verify we can't transmit any more bytes
        assert_eq!(path.mtu(1), 0);
        assert_eq!(path.mtu(10), 0);

        path.on_bytes_received(1);
        // Verify we can transmit up to 3 more bytes
        assert_eq!(path.mtu(1), 1);
        assert_eq!(path.mtu(2), 2);
        assert_eq!(path.mtu(4), 3);

        path.validated = true;
        // Validated paths should always be able to transmit
        assert_eq!(path.mtu(4), 4);
    }
}
