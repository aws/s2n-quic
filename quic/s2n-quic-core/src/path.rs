//! This module contains the Path implementation

use crate::{
    connection,
    frame::path_challenge,
    inet::SocketAddress,
    recovery::{CongestionController, RTTEstimator},
};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum State {
    Validated,
    Pending { tx_bytes: u32, rx_bytes: u32 },
}

//= https://tools.ietf.org/id/draft-ietf-quic-transport-29.txt#14.2
//# If a QUIC endpoint determines that the PMTU between any pair of local
//# and remote IP addresses has fallen below the smallest allowed maximum
//# packet size of 1200 bytes, it MUST immediately cease sending QUIC
//# packets
const MINIMUM_MTU: u16 = 1200;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Path<CC: CongestionController> {
    /// The peer's socket address
    pub peer_socket_address: SocketAddress,
    /// The connection id of the peer
    pub peer_connection_id: connection::Id,
    /// The path owns the roundtrip between peers
    pub rtt_estimator: RTTEstimator,
    /// The congestion controller for the path
    pub congestion_controller: CC,
    /// Tracks whether this path has passed Address or Path validation
    state: State,
    /// Maximum transmission unit of the path
    mtu: u16,

    //= https://tools.ietf.org/id/draft-ietf-quic-transport-29.txt#8.3
    //# To initiate path validation, an endpoint sends a PATH_CHALLENGE frame
    //# containing a random payload on the path to be validated.
    pub challenge: Option<[u8; path_challenge::DATA_LEN]>,

    /// True if the path has been validated by the peer
    peer_validated: bool,
}

/// A Path holds the local and peer socket addresses, connection ids, and state. It can be
/// validated or pending validation.
impl<CC: CongestionController> Path<CC> {
    pub fn new(
        peer_socket_address: SocketAddress,
        peer_connection_id: connection::Id,
        rtt_estimator: RTTEstimator,
        congestion_controller: CC,
        peer_validated: bool,
    ) -> Self {
        Path {
            peer_socket_address,
            peer_connection_id,
            rtt_estimator,
            congestion_controller,
            state: State::Pending {
                tx_bytes: 0,
                rx_bytes: 0,
            },
            mtu: MINIMUM_MTU,
            challenge: None,
            peer_validated,
        }
    }

    /// Called when bytes have been transmitted on this path
    pub fn on_bytes_transmitted(&mut self, bytes: usize) {
        if let State::Pending { tx_bytes, .. } = &mut self.state {
            *tx_bytes += bytes as u32;
        }
    }

    /// Called when bytes have been received on this path
    pub fn on_bytes_received(&mut self, bytes: usize) {
        if let State::Pending { rx_bytes, .. } = &mut self.state {
            *rx_bytes += bytes as u32;
        }
    }

    /// Called when the path is validated
    pub fn on_validated(&mut self) {
        self.state = State::Validated
    }

    /// Returns whether this path has passed address validation
    pub fn is_validated(&self) -> bool {
        self.state == State::Validated
    }

    /// Marks the path as peer validated
    pub fn on_peer_validated(&mut self) {
        self.peer_validated = true;
    }

    /// Returns whether this path has been validated by the peer
    pub fn is_peer_validated(&self) -> bool {
        self.peer_validated
    }

    //= https://tools.ietf.org/id/draft-ietf-quic-transport-29.txt#8.1
    //# Prior to validating the client address, servers MUST NOT send more
    //# than three times as many bytes as the number of bytes they have
    //# received.
    pub fn clamp_mtu(&self, requested_size: usize) -> usize {
        match self.state {
            State::Validated => requested_size.min(self.mtu as usize),
            State::Pending { tx_bytes, rx_bytes } => {
                let limit = rx_bytes
                    .checked_mul(3)
                    .and_then(|v| v.checked_sub(tx_bytes))
                    .unwrap_or(0);
                requested_size.min(limit as usize).min(self.mtu as usize)
            }
        }
    }

    /// Returns whether this path is blocked from transmitting more data
    pub fn at_amplification_limit(&self) -> bool {
        let mtu = self.mtu as usize;
        self.clamp_mtu(mtu) < mtu
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::recovery::congestion_controller::testing::MockCC;
    use core::time::Duration;

    #[test]
    fn amplification_limit_test() {
        let mut path = Path::new(
            SocketAddress::default(),
            connection::Id::try_from_bytes(&[]).unwrap(),
            RTTEstimator::new(Duration::from_millis(30)),
            MockCC::default(),
            true,
        );

        // Verify we enforce the amplification limit if we can't send
        // at least 1 minimum sized packet
        path.on_bytes_received(1200);
        path.on_bytes_transmitted((1200 * 2) + 1);
        assert_eq!(path.at_amplification_limit(), true);

        path.on_bytes_received(1200);
        assert_eq!(path.at_amplification_limit(), false);

        path.on_validated();
        path.on_bytes_transmitted(24);
        // Validated paths should always be able to transmit
        assert_eq!(path.at_amplification_limit(), false);
    }

    #[test]
    fn mtu_test() {
        // TODO this would work better as a fuzz test
        let mut path = Path::new(
            SocketAddress::default(),
            connection::Id::try_from_bytes(&[]).unwrap(),
            RTTEstimator::new(Duration::from_millis(30)),
            MockCC::default(),
            true,
        );

        path.on_bytes_received(3);
        path.on_bytes_transmitted(8);

        // Verify we can transmit one more byte
        assert_eq!(path.clamp_mtu(1), 1);
        assert_eq!(path.clamp_mtu(10), 1);

        path.on_bytes_transmitted(1);
        // Verify we can't transmit any more bytes
        assert_eq!(path.clamp_mtu(1), 0);
        assert_eq!(path.clamp_mtu(10), 0);

        path.on_bytes_received(1);
        // Verify we can transmit up to 3 more bytes
        assert_eq!(path.clamp_mtu(1), 1);
        assert_eq!(path.clamp_mtu(2), 2);
        assert_eq!(path.clamp_mtu(4), 3);

        path.on_validated();
        // Validated paths should always be able to transmit
        assert_eq!(path.clamp_mtu(4), 4);
    }

    #[test]
    fn peer_validated_test() {
        let mut path = Path::new(
            SocketAddress::default(),
            connection::Id::try_from_bytes(&[]).unwrap(),
            RTTEstimator::new(Duration::from_millis(30)),
            MockCC::default(),
            false,
        );

        assert!(!path.is_peer_validated());

        path.on_peer_validated();

        assert!(path.is_peer_validated());
    }
}
