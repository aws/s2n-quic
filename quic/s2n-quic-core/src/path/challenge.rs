// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    ct::ConstantTimeEq,
    frame,
    time::{Duration, Timer, Timestamp},
};

pub type Data = [u8; frame::path_challenge::DATA_LEN];

#[derive(Clone, Debug)]
pub struct State {
    retransmit_timer: Timer,
    retransmit_period: Duration,
    abandon_timer: Timer,
    data: Data,
}

#[derive(Clone, Debug)]
pub enum Challenge {
    /// There is no Challenge associated with this Path
    None,

    /// A Challenge has been sent to the peer and the response is pending
    Pending(State),

    /// A timeout caused this Challenge to be abandoned, an new Challenge will have to be used
    Abandoned,
}

impl Challenge {
    pub fn new(
        timestamp: Timestamp,
        retransmit_period: Duration,
        expiration: Duration,
        data: Data,
    ) -> Self {
        let mut retransmit_timer = Timer::default();
        retransmit_timer.set(timestamp);
        let mut abandon_timer = Timer::default();
        abandon_timer.set(timestamp + expiration);

        let state = State {
            retransmit_timer,
            retransmit_period,
            abandon_timer,
            data,
        };

        Self::Pending(state)
    }

    pub fn timers(&self) -> impl Iterator<Item = Timestamp> {
        if let Challenge::Pending(state) = self {
            Some(
                core::iter::empty()
                    .chain(state.abandon_timer.iter())
                    .chain(state.retransmit_timer.iter()),
            )
        } else {
            None
        }
        .into_iter()
        .flatten()
    }

    /// When a PATH_CHALLENGE is transmitted this handles any internal state operations.
    pub fn on_transmit(&mut self, timestamp: Timestamp) {
        if let Challenge::Pending(state) = self {
            state
                .retransmit_timer
                .set(timestamp + state.retransmit_period);
        }
    }

    pub fn on_timeout(&mut self, timestamp: Timestamp) {
        if let Challenge::Pending(state) = self {
            if state.abandon_timer.is_expired(timestamp) {
                *self = Challenge::Abandoned;
            }
        }
    }

    pub fn is_pending(&self, timestamp: Timestamp) -> bool {
        if let Challenge::Pending(state) = self {
            return state.retransmit_timer.is_expired(timestamp);
        }

        false
    }

    pub fn data(&self) -> Option<&Data> {
        if let Challenge::Pending(state) = self {
            return Some(&state.data);
        }
        None
    }

    pub fn is_valid(&self, timestamp: Timestamp, data: &[u8]) -> bool {
        if let Challenge::Pending(state) = self {
            let mut valid = true;
            if state.abandon_timer.is_expired(timestamp) {
                valid = false;
            }

            if ConstantTimeEq::ct_eq(&state.data[..], &data).unwrap_u8() == 0 {
                valid = false;
            }

            return valid;
        }

        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::time::{Clock, Duration, NoopClock};

    #[test]
    fn test_challenge_validation() {
        let clock = NoopClock {};
        let expiration = Duration::from_millis(100);
        let expected_data: [u8; 8] = [0; 8];

        let challenge = Challenge::new(
            clock.get_time(),
            Duration::from_millis(0),
            expiration,
            expected_data,
        );

        assert!(challenge.is_valid(clock.get_time(), &expected_data));

        let empty_challenge = Challenge::None;
        assert!(!empty_challenge.is_valid(clock.get_time(), &expected_data));

        let invalid_data = [1; 8];
        assert!(!challenge.is_valid(clock.get_time(), &invalid_data));

        let expired = Duration::from_millis(150);
        assert!(!challenge.is_valid(clock.get_time() + expired, &expected_data));
    }
}
