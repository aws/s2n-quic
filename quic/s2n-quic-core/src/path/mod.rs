// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    event,
    inet::{IpV4Address, IpV6Address, SocketAddress, SocketAddressV4, SocketAddressV6},
};
use core::fmt;

#[cfg(any(test, feature = "generator"))]
use bolero_generator::*;

pub mod ecn;
pub mod migration;
pub mod mtu;

pub use mtu::*;

// Initial PTO backoff multiplier is 1 indicating no additional increase to the backoff.
pub const INITIAL_PTO_BACKOFF: u32 = 1;

/// Internal Id of a path in the manager
#[derive(Clone, Copy, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Id(u8);

impl Id {
    /// Create a new path::Id
    ///
    /// # Safety
    /// This should only be used by the path::Manager
    pub unsafe fn new(id: u8) -> Self {
        Self(id)
    }

    pub fn as_u8(&self) -> u8 {
        self.0
    }
}

impl event::IntoEvent<u64> for Id {
    #[inline]
    fn into_event(self) -> u64 {
        self.0 as u64
    }
}

#[cfg(any(test, feature = "testing"))]
impl Id {
    pub fn test_id() -> Self {
        unsafe { Id::new(0) }
    }
}

/// An interface for an object that represents a unique path between two endpoints
pub trait Handle: 'static + Copy + Send + fmt::Debug {
    /// Creates a Handle from a RemoteAddress
    fn from_remote_address(remote_addr: RemoteAddress) -> Self;

    /// Returns the remote address for the given handle
    fn remote_address(&self) -> RemoteAddress;

    /// Updates the remote port to the given value
    fn set_remote_port(&mut self, port: u16);

    /// Returns the local address for the given handle
    fn local_address(&self) -> LocalAddress;

    /// Returns `true` if the two handles are equal from a network perspective
    ///
    /// This function is used to determine if a connection has migrated to another
    /// path.
    fn eq(&self, other: &Self) -> bool;

    /// Returns `true` if the two handles are strictly equal to each other, i.e.
    /// byte-for-byte.
    fn strict_eq(&self, other: &Self) -> bool;

    /// Depending on the current value of `self`, fields from `other` may be copied to increase the
    /// fidelity of the value.
    ///
    /// This is especially useful for clients that initiate a connection only based on the remote
    /// IP and port. They likely wouldn't know the IP address of the local socket. Once a response
    /// is received from the server, the IP information will be known at this point and the handle
    /// can be updated with the new information.
    ///
    /// Implementations should try to limit the cost of updating by checking the current value to
    /// see if it needs updating.
    fn maybe_update(&mut self, other: &Self);
}

macro_rules! impl_addr {
    ($name:ident) => {
        #[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Hash)]
        #[cfg_attr(any(test, feature = "generator"), derive(TypeGenerator))]
        #[cfg_attr(kani, derive(kani::Arbitrary))]
        pub struct $name(pub SocketAddress);

        impl From<event::api::SocketAddress<'_>> for $name {
            #[inline]
            fn from(value: event::api::SocketAddress<'_>) -> Self {
                match value {
                    event::api::SocketAddress::IpV4 { ip, port } => {
                        $name(IpV4Address::new(*ip).with_port(port).into())
                    }
                    event::api::SocketAddress::IpV6 { ip, port } => {
                        $name(IpV6Address::new(*ip).with_port(port).into())
                    }
                }
            }
        }

        impl From<SocketAddress> for $name {
            #[inline]
            fn from(value: SocketAddress) -> Self {
                Self(value)
            }
        }

        impl From<SocketAddressV4> for $name {
            #[inline]
            fn from(value: SocketAddressV4) -> Self {
                Self(value.into())
            }
        }

        impl From<SocketAddressV6> for $name {
            #[inline]
            fn from(value: SocketAddressV6) -> Self {
                Self(value.into())
            }
        }

        impl core::ops::Deref for $name {
            type Target = SocketAddress;

            fn deref(&self) -> &Self::Target {
                &self.0
            }
        }

        impl core::ops::DerefMut for $name {
            fn deref_mut(&mut self) -> &mut Self::Target {
                &mut self.0
            }
        }
    };
}

impl_addr!(LocalAddress);

impl_addr!(RemoteAddress);

impl Handle for RemoteAddress {
    #[inline]
    fn from_remote_address(remote_address: RemoteAddress) -> Self {
        remote_address
    }

    #[inline]
    fn remote_address(&self) -> RemoteAddress {
        *self
    }

    #[inline]
    fn set_remote_port(&mut self, port: u16) {
        self.0.set_port(port)
    }

    #[inline]
    fn local_address(&self) -> LocalAddress {
        SocketAddressV4::UNSPECIFIED.into()
    }

    #[inline]
    fn eq(&self, other: &Self) -> bool {
        PartialEq::eq(&self.unmap(), &other.unmap())
    }

    #[inline]
    fn strict_eq(&self, other: &Self) -> bool {
        PartialEq::eq(self, other)
    }

    #[inline]
    fn maybe_update(&mut self, _other: &Self) {
        // nothing to update
    }
}

#[derive(Clone, Copy, Debug, Eq)]
#[cfg_attr(any(test, feature = "generator"), derive(TypeGenerator))]
pub struct Tuple {
    pub remote_address: RemoteAddress,
    pub local_address: LocalAddress,
}

impl PartialEq for Tuple {
    #[inline]
    fn eq(&self, other: &Self) -> bool {
        PartialEq::eq(&self.remote_address, &other.remote_address)
            && PartialEq::eq(&self.local_address, &other.local_address)
    }
}

impl Handle for Tuple {
    #[inline]
    fn from_remote_address(remote_address: RemoteAddress) -> Self {
        let local_address = SocketAddressV4::UNSPECIFIED.into();
        Self {
            remote_address,
            local_address,
        }
    }

    #[inline]
    fn remote_address(&self) -> RemoteAddress {
        self.remote_address
    }

    #[inline]
    fn set_remote_port(&mut self, port: u16) {
        self.remote_address.set_port(port)
    }

    #[inline]
    fn local_address(&self) -> LocalAddress {
        self.local_address
    }

    #[inline]
    fn eq(&self, other: &Self) -> bool {
        PartialEq::eq(&self.local_address.unmap(), &other.local_address.unmap())
            && Handle::eq(&self.remote_address, &other.remote_address)
    }

    #[inline]
    fn strict_eq(&self, other: &Self) -> bool {
        PartialEq::eq(self, other)
    }

    #[inline]
    fn maybe_update(&mut self, other: &Self) {
        if other.local_address.port() == 0 {
            return;
        }

        // once we discover our path, or the port changes, update the address with the new information
        if self.local_address.port() != other.local_address.port() {
            self.local_address = other.local_address;
        }
    }
}

//= https://www.rfc-editor.org/rfc/rfc9308#section-8.1
//# Some UDP protocols are vulnerable to reflection attacks, where an
//# attacker is able to direct traffic to a third party as a denial of
//# service.  For example, these source ports are associated with
//# applications known to be vulnerable to reflection attacks, often due
//# to server misconfiguration:
//#
//# *  port 53 - DNS [RFC1034]
//#
//# *  port 123 - NTP [RFC5905]
//#
//# *  port 1900 - SSDP [SSDP]
//#
//# *  port 5353 - mDNS [RFC6762]
//#
//# *  port 11211 - memcache

/// List of ports to refuse connections from. This list must be sorted.
///
/// Based on https://quiche.googlesource.com/quiche/+/bac04054bccb2a249d4705ecc94a646404d41c1b/quiche/quic/core/quic_dispatcher.cc#498
const BLOCKED_PORTS: [u16; 11] = [
    0,   // We cannot send to port 0 so drop that source port.
    17,  // Quote of the Day, can loop with QUIC.
    19,  // Chargen, can loop with QUIC.
    53,  // DNS, vulnerable to reflection attacks.
    111, // Portmap.
    123, // NTP, vulnerable to reflection attacks.
    137, // NETBIOS Name Service,
    138, // NETBIOS Datagram Service
    161, // SNMP.
    389, // CLDAP.
    500, // IKE, can loop with QUIC.
];
const MAX_BLOCKED_PORT: u16 = BLOCKED_PORTS[BLOCKED_PORTS.len() - 1];

#[inline]
pub fn remote_port_blocked(port: u16) -> bool {
    if port > MAX_BLOCKED_PORT {
        // Early return to avoid iteration if possible
        return false;
    }

    for blocked in BLOCKED_PORTS {
        if port == blocked {
            return true;
        }
    }

    false
}

// The below ports are also vulnerable to reflection attacks, but are within
// the original ephemeral port range of 1024â€“65535, so there is a chance
// clients may be randomly assigned them. To address this, we can throttle
// connections using these ports instead of fully blocking them.

/// List of ports to throttle connections from. This list must be sorted.
///
/// Based on https://quiche.googlesource.com/quiche/+/bac04054bccb2a249d4705ecc94a646404d41c1b/quiche/quic/core/quic_dispatcher.cc#498
const THROTTLED_PORTS: [u16; 5] = [
    1900,  // SSDP, vulnerable to reflection attacks.
    3702,  // WS-Discovery, vulnerable to reflection attacks.
    5353,  // mDNS, vulnerable to reflection attacks.
    5355,  // LLMNR, vulnerable to reflection attacks.
    11211, // memcache, vulnerable to reflection attacks.
];
const MAX_THROTTLED_PORT: u16 = THROTTLED_PORTS[THROTTLED_PORTS.len() - 1];
pub const THROTTLED_PORTS_LEN: usize = THROTTLED_PORTS.len();

#[inline]
pub fn remote_port_throttled_index(port: u16) -> Option<usize> {
    for (idx, throttled_port) in THROTTLED_PORTS.iter().enumerate() {
        if *throttled_port == port {
            return Some(idx);
        }
    }
    None
}

#[inline]
pub fn remote_port_throttled(port: u16) -> bool {
    if port > MAX_THROTTLED_PORT {
        // Early return to avoid iteration if possible
        return false;
    }

    for throttled in THROTTLED_PORTS {
        if port == throttled {
            return true;
        }
    }

    false
}

#[cfg(test)]
mod tests {
    use crate::path::{
        remote_port_blocked, remote_port_throttled, BLOCKED_PORTS, MAX_BLOCKED_PORT,
        MAX_THROTTLED_PORT, THROTTLED_PORTS,
    };

    #[test]
    fn blocked_ports_is_sorted() {
        assert_eq!(Some(MAX_BLOCKED_PORT), BLOCKED_PORTS.iter().max().copied());

        let mut sorted = BLOCKED_PORTS.to_vec();
        sorted.sort_unstable();

        for i in 0..BLOCKED_PORTS.len() {
            assert_eq!(sorted[i], BLOCKED_PORTS[i]);
        }
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn blocked_port() {
        for port in 0..u16::MAX {
            let blocked_expected = BLOCKED_PORTS.iter().copied().any(|blocked| blocked == port);
            assert_eq!(blocked_expected, remote_port_blocked(port));
        }
    }

    #[test]
    fn throttled_ports_is_sorted() {
        assert_eq!(
            Some(MAX_THROTTLED_PORT),
            THROTTLED_PORTS.iter().max().copied()
        );

        let mut sorted = THROTTLED_PORTS.to_vec();
        sorted.sort_unstable();

        for i in 0..THROTTLED_PORTS.len() {
            assert_eq!(sorted[i], THROTTLED_PORTS[i]);
        }
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn throttled_port() {
        for port in 0..u16::MAX {
            let throttled_expected = THROTTLED_PORTS
                .iter()
                .copied()
                .any(|throttled| throttled == port);
            assert_eq!(throttled_expected, remote_port_throttled(port));
        }
    }
}

#[cfg(any(test, feature = "testing"))]
pub mod testing {
    use crate::{
        connection, event,
        event::{builder::SocketAddress, IntoEvent},
    };

    impl<'a> event::builder::Path<'a> {
        pub fn test() -> Self {
            Self {
                local_addr: SocketAddress::IpV4 {
                    ip: &[127, 0, 0, 1],
                    port: 0,
                },
                local_cid: connection::LocalId::TEST_ID.into_event(),
                remote_addr: SocketAddress::IpV4 {
                    ip: &[127, 0, 0, 1],
                    port: 0,
                },
                remote_cid: connection::PeerId::TEST_ID.into_event(),
                id: 0,
                is_active: false,
            }
        }
    }
}
