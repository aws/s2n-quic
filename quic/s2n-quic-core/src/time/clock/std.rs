// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;
use ::std::time::Instant;

impl<C: 'static + Clock> Clock for &'static ::std::thread::LocalKey<C> {
    fn get_time(&self) -> Timestamp {
        self.with(|clock| clock.get_time())
    }
}

#[derive(Clone, Copy, Debug)]
pub struct StdClock {
    epoch: Instant,
}

impl Default for StdClock {
    fn default() -> Self {
        Self {
            epoch: Instant::now(),
        }
    }
}

impl StdClock {
    /// Creates a new `StdClock` with the given epoch
    pub const fn new(epoch: Instant) -> Self {
        Self { epoch }
    }
}

impl Clock for StdClock {
    fn get_time(&self) -> Timestamp {
        unsafe { Timestamp::from_duration(self.epoch.elapsed()) }
    }
}

#[test]
#[cfg_attr(miri, ignore)] // time isn't queryable in miri
fn monotonicity_test() {
    let clock = StdClock::default();
    let ts1 = clock.get_time();
    ::std::thread::sleep(Duration::from_millis(50));
    let ts2 = clock.get_time();
    assert!(ts2 - ts1 >= Duration::from_millis(50));
}
