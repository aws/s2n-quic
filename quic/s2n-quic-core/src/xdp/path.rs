// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    inet::{ethernet::MacAddress, ipv4, IpAddress},
    path::{self, Handle},
};

#[cfg(any(test, feature = "generator"))]
use bolero_generator::prelude::*;

macro_rules! define_address {
    ($name:ident) => {
        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
        #[cfg_attr(any(test, feature = "generator"), derive(TypeGenerator))]
        pub struct $name {
            pub mac: MacAddress,
            pub ip: IpAddress,
            pub port: u16,
        }

        impl $name {
            pub const UNSPECIFIED: Self = Self {
                mac: MacAddress::UNSPECIFIED,
                ip: IpAddress::Ipv4(ipv4::IpV4Address::UNSPECIFIED),
                port: 0,
            };

            #[inline]
            pub fn unmap(self) -> Self {
                Self {
                    mac: self.mac,
                    ip: self.ip.unmap(),
                    port: self.port,
                }
            }
        }

        impl From<path::$name> for $name {
            #[inline]
            fn from(addr: path::$name) -> Self {
                Self {
                    mac: MacAddress::UNSPECIFIED,
                    ip: addr.ip(),
                    port: addr.port(),
                }
            }
        }

        impl From<$name> for path::$name {
            #[inline]
            fn from(addr: $name) -> Self {
                addr.ip.with_port(addr.port).into()
            }
        }
    };
}

define_address!(RemoteAddress);
define_address!(LocalAddress);

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "generator"), derive(TypeGenerator))]
pub struct Tuple {
    pub remote_address: RemoteAddress,
    pub local_address: LocalAddress,
}

impl Tuple {
    pub const UNSPECIFIED: Self = Self {
        remote_address: RemoteAddress::UNSPECIFIED,
        local_address: LocalAddress::UNSPECIFIED,
    };

    #[inline]
    pub fn swap(&mut self) {
        core::mem::swap(&mut self.remote_address.mac, &mut self.local_address.mac);
        core::mem::swap(&mut self.remote_address.ip, &mut self.local_address.ip);
        core::mem::swap(&mut self.remote_address.port, &mut self.local_address.port);
    }
}

impl Handle for Tuple {
    #[inline]
    fn from_remote_address(remote_address: path::RemoteAddress) -> Self {
        let remote_address = remote_address.into();
        let local_address = LocalAddress::UNSPECIFIED;
        Self {
            remote_address,
            local_address,
        }
    }

    #[inline]
    fn remote_address(&self) -> path::RemoteAddress {
        self.remote_address.into()
    }

    #[inline]
    fn set_remote_port(&mut self, port: u16) {
        self.remote_address.port = port;
    }

    #[inline]
    fn local_address(&self) -> path::LocalAddress {
        self.local_address.into()
    }

    #[inline]
    fn eq(&self, other: &Self) -> bool {
        // TODO only compare everything if the other is all filled out
        PartialEq::eq(&self.local_address.unmap(), &other.local_address.unmap())
            && PartialEq::eq(&self.remote_address.unmap(), &other.remote_address.unmap())
    }

    #[inline]
    fn strict_eq(&self, other: &Self) -> bool {
        PartialEq::eq(self, other)
    }

    #[inline]
    fn maybe_update(&mut self, other: &Self) {
        if other.local_address.port == 0 {
            return;
        }

        // once we discover our path, or the port changes, update the address with the new information
        if self.local_address.port != other.local_address.port {
            *self = *other;
        }
    }
}
