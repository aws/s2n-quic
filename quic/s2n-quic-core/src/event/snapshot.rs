// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use core::panic;
use std::path::{Path, PathBuf};

#[derive(Clone, Debug)]
pub struct Location {
    location: &'static panic::Location<'static>,
    name: String,
}

impl Location {
    #[track_caller]
    pub fn new<N: core::fmt::Display>(name: N) -> Self {
        let location = panic::Location::caller();
        let name = name.to_string();
        Self { location, name }
    }

    #[track_caller]
    #[allow(clippy::manual_map)] // using `Option::map` messes up the track_caller
    pub fn from_thread_name() -> Option<Self> {
        let thread = std::thread::current();

        // only create a location if insta can figure out the test name from the
        // thread
        if let Some(name) = thread.name().filter(|name| *name != "main") {
            let name = name
                .split("::")
                .chain(Some("events"))
                .collect::<Vec<_>>()
                .join("__");
            Some(Self::new(name))
        } else {
            None
        }
    }

    pub fn snapshot_log(&self, output: &[String]) {
        // miri doesn't support the syscalls that insta uses
        if cfg!(miri) {
            return;
        }

        let value = output.join("\n");

        let name = self.name.as_str();

        let mut settings = insta::Settings::clone_current();

        // we want to use the actual caller's module
        settings.set_prepend_module_to_snapshot(false);
        settings.set_input_file(self.location.file());
        settings.set_snapshot_path(self.snapshot_path());
        settings.set_omit_expression(true);

        settings.bind(|| {
            insta::assert_snapshot!(name, &value);
        });
    }

    fn snapshot_path(&self) -> PathBuf {
        let ws = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/../.."));
        let file = Path::new(self.location.file());

        let file = if file.is_relative() {
            ws.join(file)
        } else {
            file.to_path_buf()
        };

        file.canonicalize()
            .unwrap()
            .parent()
            .unwrap()
            .join("snapshots")
    }
}
