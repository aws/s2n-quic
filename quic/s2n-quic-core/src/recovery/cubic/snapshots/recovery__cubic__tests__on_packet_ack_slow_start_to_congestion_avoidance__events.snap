---
source: quic/s2n-quic-core/src/event/snapshot.rs
input_file: quic/s2n-quic-core/src/recovery/cubic/tests.rs
---
SlowStartExited { path_id: 0, cause: Other, congestion_window: 10100 }
